hal/mpu/mbed_mpu_v7m.c

```diff
@@ -0,0 +1,156 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "hal/mpu_api.h"
+#include "cmsis.h"
+
+#if ((__ARM_ARCH_7M__ == 1U) || (__ARM_ARCH_7EM__ == 1U) || (__ARM_ARCH_6M__ == 1U)) && \
+    defined (__MPU_PRESENT) && (__MPU_PRESENT == 1U)
+
+#if !DEVICE_MPU
+#error "Device has v7m MPU but it is not enabled. Add 'MPU' to device_has in targets.json"
+#endif
+
+void mbed_mpu_init()
+{
+    mbed_mpu_enable_ram_xn(false);
+}
+
+void mbed_mpu_free()
+{
+    mbed_mpu_enable_ram_xn(false);
+}
+
+void mbed_mpu_enable_ram_xn(bool enable)
+{
+    // Flush memory writes before configuring the MPU.
+    __DSB();
+
+    const uint32_t regions = (MPU->TYPE & MPU_TYPE_DREGION_Msk) >> MPU_TYPE_DREGION_Pos;
+
+    // Disable the MCU
+    MPU->CTRL = 0;
+
+    // Reset all mapping
+    for (uint32_t i = 0; i < regions; i++) {
+        ARM_MPU_ClrRegion(i);
+    }
+
+    if (!enable) {
+        return;
+    }
+
+    /*
+     * ARMv6m and ARMv7m memory map:
+     *
+     * Start        End            Name            Executable by default        Mbed MPU protection
+     * 0x00000000 - 0x1FFFFFFF     Code            Yes                          Execute disabled for second half
+     * 0x20000000 - 0x3FFFFFFF     SRAM            Yes                          Execute disabled
+     * 0x40000000 - 0x5FFFFFFF     Peripheral      No
+     * 0x60000000 - 0x7FFFFFFF     RAM             Yes                          Execute disabled
+     * 0x80000000 - 0x9FFFFFFF     RAM             Yes                          Execute disabled
+     * 0xA0000000 - 0xBFFFFFFF     Device          No
+     * 0xC0000000 - 0xDFFFFFFF     Device          No
+     * 0xE0000000 - 0xFFFFFFFF     System          No
+     */
+
+    if (regions >= 3) {
```

Comment:
As I know of no devices with >=3 and < 8 regions, leaving 5 regions unused and unusable seems wasteful. I would **really** like the ROM to be marked read-only to give us precise abort on NULL writes. That would take us to 4 regions, and on current devices would leave 4 available to system code if we didn't overwrite them.

(Would presumably need another API like the "disable XN" to mark ROM writable for flashing)